skypy/halo/mass.py

```diff
@@ -1,38 +1,245 @@
-r'''Halo mass module.
-
-This module provides methods to sample the  masses of dark matter halos.
+r'''Halo mass sampler.
+This code samples halos from their mass function.
 
 Models
 ======
-
 .. autosummary::
    :nosignatures:
    :toctree: ../api/
 
    press_schechter
-
+   halo_mass_function
+   sheth_tormen_collapse_function
+   press_schechter_collapse_function
 '''
 
 import numpy as np
+from scipy import integrate
+from functools import partial
+from astropy.constants import G
+from astropy import units as u
 
+from skypy.power_spectrum import growth_function
 from skypy.utils.random import schechter
 
+__all__ = [
+    'press_schechter',
+    'halo_mass_function',
+    'sheth_tormen_collapse_function',
+    'press_schechter_collapse_function',
+ ]
+
+
+def _sigma_squared(M, wavenumber, linear_power_today, cosmology, redshift):
+    k = wavenumber
+    Pk = linear_power_today
+    Hz = cosmology.H(redshift).value
+    if isinstance(M, np.ndarray):
+        M = M[:, np.newaxis]
+
+    # The linear frowth function
+    D0 = growth_function(0, cosmology)
+    Dz = growth_function(redshift, cosmology) / D0
+
+    R = np.power(2 * G.value * M / Hz, 1.0 / 3.0)
+    j1 = (np.sin(k * R) - np.cos(k * R) * k * R) / (k * R)
+    top_hat = 3 * j1 / (k * R)**2
+    integrand = Pk * top_hat**2.0 * k**2.0
+    return Dz**2 * integrate.simps(integrand, k) / (2. * np.pi ** 2.)
+
+
+_delta_critical_squared = partial(_sigma_squared, redshift=0)
+
+
+def halo_mass_function(collapse_function, m_min, m_max, m_star, redshift,
```

Comment:
I think most users would be much more familiar with, and prefer, to work in terms of a critical density (which should have a default value 1.686) than m_star.